src/wasi.rs

```diff
@@ -143,98 +136,53 @@ impl Wasi {
         self.inner
             .map_dir(alias.as_str(), PathBuf::from(directory.to_string()))
-            .map_err(|error| {
-                RuntimeError::py_err(format!(
-                    "Failed to configure the map directory when creating the WASI state: {}",
-                    error
-                ))
-            })?;
+            .map_err(to_py_err::<RuntimeError, _>)?;
 
         Ok(())
     }
 }
 
 #[pymethods]
-impl Wasi {
-    /// Build a `Wasi` object. The constructor can be used to
-    /// initialize its state, and its methods help to update its
-    /// state.
-    ///
-    /// # Examples
-    ///
-    /// Thus, both next notations are equivalent and can be mixed:
-    ///
-    /// ```py
-    /// wasi = Wasi(
-    ///     program_name="wasi_test_program",
-    ///     arguments=["--test"],
-    ///     environments={"COLOR": "true", "APP_SHOULD_LOG": "false"},
-    ///     map_directories={"the_host_current_dir": "."}
-    /// )
-    /// ```
-    ///
-    /// could be rewritten:
-    ///
-    /// ```py
-    /// wasi = \
-    ///     Wasi("wasi_test_program"). \
-    ///         argument("--test"). \
-    ///         environment("COLOR", "true"). \
-    ///         environment("APP_SHOULD_LOG", "false"). \
-    ///         map_directory("the_host_current_dir", ".")
-    /// ```
+impl StateBuilder {
     #[new]
-    #[args(
-        arguments = "PyList::empty(_py)",
-        environments = "PyDict::new(_py)",
-        preopen_directories = "PyList::empty(_py)",
-        map_directories = "PyDict::new(_py)"
-    )]
     fn new(
         program_name: String,
-        arguments: &PyList,
-        environments: &PyDict,
-        preopen_directories: &PyList,
-        map_directories: &PyDict,
+        arguments: Option<&PyList>,
+        environments: Option<&PyDict>,
+        preopen_directories: Option<&PyList>,
+        map_directories: Option<&PyDict>,
     ) -> PyResult<Self> {
         let mut wasi = Self {
-            inner: state::WasiState::new(program_name.as_str()),
+            inner: wasmer_wasi::WasiState::new(program_name.as_str()),
         };
 
-        if !arguments.is_empty() {
-            wasi.self_arguments(arguments);
+        if arguments.is_some() {
+            wasi.self_arguments(arguments.unwrap());
         }
 
-        if !environments.is_empty() {
-            wasi.self_environments(environments);
+        if environments.is_some() {
+            wasi.self_environments(environments.unwrap());
         }
 
-        if !preopen_directories.is_empty() {
-            wasi.self_preopen_directories(preopen_directories)?;
+        if preopen_directories.is_some() {
+            wasi.self_preopen_directories(preopen_directories.unwrap())?;
         }
 
-        if !map_directories.is_empty() {
-            wasi.self_map_directories(map_directories)?;
+        if map_directories.is_some() {
+            wasi.self_map_directories(map_directories.unwrap())?;
```

Comment:
I think `if let Some(v) = thing` will be a bit cleaner here and avoids `unwrap`

Reply:
`Option` has a method called `take` if you want to take ownership

Reply:
Of course, `if let Some(v) = thing` 🤦. It was a “temporary” update of the code. Didn't notice that. Updating!